internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,308 @@
+// Each reconciliation event gets the associated object from Kubernetes before passing it to Reconcile
+// Notes for test, this function expect req to come in with k8s UID already populated like real cluster
+func (r *NonAdminDownloadRequestReconciler) Reconcile(ctx context.Context, req *nacv1alpha1.NonAdminDownloadRequest) (reconcile.Result, error) {
+	if req == nil || req.Spec.Target.Kind == constant.EmptyString {
+		return ctrl.Result{}, nil
+	}
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling NonAdminDownloadRequest", "name", req.Name, "namespace", req.Namespace)
+	// defines associated downloadrequest for getting, or deleting
+	veleroDR := velerov1.DownloadRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.VeleroDownloadRequestName(),
+			Namespace: r.OADPNamespace,
+			Labels: func() map[string]string {
+				nal := function.GetNonAdminLabels()
+				nal[constant.NadrOriginNACUUIDLabel] = string(req.GetUID())
+				return nal
+			}(),
+			Annotations: function.GetNonAdminDownloadRequestAnnotations(req),
+		},
+		Spec: velerov1.DownloadRequestSpec{
+			Target: velerov1.DownloadTarget{
+				Kind: req.Spec.Target.Kind,
+			},
+		},
+	}
+	// if request has status, it may already be processed or expired
+	if req.Status.VeleroDownloadRequest.Status != nil {
+		// if request is expired, delete NADR after deleting velero DR
+		if req.Status.VeleroDownloadRequest.Status.Expiration != nil && req.Status.VeleroDownloadRequest.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+			// find associated velero downloadrequest and delete that first
+			return reconcile.Result{}, r.deleteExpiredDownloadRequest(ctx, &veleroDR, req)
+		}
+		// if request is not expired, and has downloadUrl, requeue when expired to cleanup.
+		if req.Status.VeleroDownloadRequest.Status.DownloadURL != constant.EmptyString {
+			return reconcile.Result{RequeueAfter: time.Until(req.Status.VeleroDownloadRequest.Status.Expiration.Time)}, nil
+		}
+	}
+	// try get veleroDR if exists, then update status
+	if err := r.Get(ctx, types.NamespacedName{Namespace: veleroDR.Namespace, Name: veleroDR.Name}, &veleroDR); err == nil {
+		return reconcile.Result{}, r.updateNADRWithURL(ctx, &veleroDR, req)
+	} else if !apierrors.IsNotFound(err) {
+		// some other errors, requeue to retry get
+		return reconcile.Result{}, err
+	}
+	// veleroDR is not found, so we create one
+	return reconcile.Result{}, r.processDownloadRequest(ctx, &veleroDR, req)
+}
+
+// Delete expired download request
+// find associated velero downloadrequest and delete that first
+func (r *NonAdminDownloadRequestReconciler) deleteExpiredDownloadRequest(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("Deleting expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), req.Namespace)
+	if err := r.Delete(ctx, veleroDR); err != nil {
```

Comment:
is not this Velero responsibility?

Reply:
No. Velero does not garbage collect.

I think they pile up.

https://github.com/vmware-tanzu/velero/pull/5329

Velero might implement gc eventually until then our logic here is fine imo.

Reply:
is this not garbage collection for it?

https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/download_request_controller.go#L108

Reply:
we can let velero delete, and on velero delete event we can delete nadr.

Reply:
what do we want to do with things without expiration? ie. velero server too busy to handle and process.. no url therefore no expiration.

Reply:
shubham said GC on NADR without a corresponding velero DR.. I assume we want to only do this if NADR.status.url is populated which is already processed.

Reply:
velero once a minute reconcile and check for expiration. we can defer to velero for a successful case.
we need to handle case without expiration due to errors possibly but future enhancement/discussion.